src/report/bulk.rs

```diff
@@ -0,0 +1,202 @@
+//! constraints
+//!
+use crate::compact_log::CompactLogItem;
+use crate::history::DynamicHistory;
+use crate::{ExtensionBytes, ReportError, TracerId};
+use core::mem::{align_of, size_of};
+use static_assertions::{assert_eq_align, const_assert_eq};
+
+/// Write reports all at once into a single byte slice.
+pub trait BulkReporter {
+    /// Conduct necessary background activities and write
+    /// the recorded reporting log to a collection backend,
+    /// including arbitrary extension bytes provided.
+    ///
+    /// Writes the Tracer's internal state according to the
+    /// log reporting schema.
+    ///
+    /// If the write was successful, returns the number of *bytes* written
+    fn report_with_extension(
+        &mut self,
+        destination: &mut [u8],
+        extension_metadata: ExtensionBytes<'_>,
+    ) -> Result<usize, ReportError>;
+
+    /// Conduct necessary background activities and write
+    /// the recorded reporting log to a collection backend.
+    ///
+    /// Writes the Tracer's internal state according to the
+    /// bulk log reporting format.
+    ///
+    /// If the write was successful, returns the number of *bytes* written
+    fn report(&mut self, destination: &mut [u8]) -> Result<usize, ReportError> {
+        self.report_with_extension(destination, ExtensionBytes(&[]))
+    }
+}
+impl<'log> BulkReporter for BulkReportSourceComponents<'log> {
+    fn report_with_extension(
+        &mut self,
+        destination: &mut [u8],
+        extension_metadata: ExtensionBytes<'_>,
+    ) -> Result<usize, ReportError> {
+        let n_log_bytes = self.log.len() * size_of::<CompactLogItem>();
+        let n_extension_bytes = extension_metadata.0.len();
+        if n_log_bytes > core::u32::MAX as usize {
+            // We don't currently support sending reports with logs that big.
+            // N.B. Point for future improvement - actually only send some of the log
+            // by doing some pre-segmentation and determining where we can cut it up,
+            // possibly injecting duplicate clocks as necessary.
+            return Err(ReportError::Encoding);
+        }
+        if n_extension_bytes > core::u32::MAX as usize {
+            // We don't support sending reports with extensions that big.
+            // There likely isn't a valid remediation available
+            // since the extension data is opaque.
+            return Err(ReportError::Extension);
+        }
+        let required_bytes = size_of::<WireBulkHeader>() + n_log_bytes + n_extension_bytes;
+        if destination.len() < required_bytes {
+            return Err(ReportError::InsufficientDestinationSize);
+        }
+        // We consider this relatively safe because WireBulkHeader is a largely
+        // uninterpreted pile of bytes fields all with alignments == 1
+        // and we know from the above check that there's enough size
+        assert_eq_align!(u8, WireBulkHeader);
```

Comment:
This is confusing to me. Isn't /everything/ byte-aligned?

Reply:
Of is this asserting that WireBulkHeader doesn't require some coarser alignment?

Reply:
Sanity-check of the latter.